nixos/modules/services/databases/openldap.nix

```diff
@@ -244,57 +356,156 @@ in
+        } // (lib.optionalAttrs (cfg.suffix != null) {
+          olcSuffix = cfg.suffix;
+        }) // (lib.optionalAttrs (cfg.dataDir != null) {
+          olcDbDirectory = cfg.dataDir;
+        }) // (lib.optionalAttrs (cfg.rootdn != null) {
+          olcRootDN = cfg.rootdn; # TODO: Optional
+        }) // (lib.optionalAttrs (cfg.rootpw != null || cfg.rootpwFile != null) {
+          olcRootPW = (if cfg.rootpwFile != null then { path = cfg.rootpwFile; } else cfg.rootpw); # TODO: Optional
+        });
+      });
+    };
+
     systemd.services.openldap = {
       description = "LDAP server";
       wantedBy = [ "multi-user.target" ];
       after = [ "network.target" ];
-      preStart = ''
+      preStart = let
+        dbSettings = lib.filterAttrs (name: value: lib.hasPrefix "olcDatabase=" name) cfg.settings.children;
+        dataDirs = lib.mapAttrsToList (name: value: value.attrs.olcDbDirectory) dbSettings;
+        settingsFile = pkgs.writeText "config.ldif" (lib.concatStringsSep "\n" (attrsToLdif "cn=config" cfg.settings));
+      in ''
         mkdir -p /run/slapd
         chown -R "${cfg.user}:${cfg.group}" /run/slapd
+
+        mkdir -p '${configDir}' ${lib.escapeShellArgs dataDirs}
+        chown "${cfg.user}:${cfg.group}" '${configDir}' ${lib.escapeShellArgs dataDirs}
+
+        ${lib.optionalString (cfg.configDir == null) (
+          if (cfg.extraConfig != "" || cfg.extraDatabaseConfig != "") then ''
+            rm -Rf '${configDir}'/*
+            # -u disables config generation, so just ignore the return code
+            ${openldap}/bin/slaptest -f ${configFile} -F ${configDir} || true
+          '' else ''
+            rm -Rf '${configDir}'/*
+            ${openldap}/bin/slapadd -F ${configDir} -n0 -l ${settingsFile}
+          ''
+        )}
+        chown -R "${cfg.user}:${cfg.group}" '${configDir}'
+
         ${optionalString (cfg.declarativeContents != null) ''
-          rm -Rf "${cfg.dataDir}"
-        ''}
-        mkdir -p "${cfg.dataDir}"
-        ${optionalString (cfg.declarativeContents != null) ''
-          ${openldap.out}/bin/slapadd ${configOpts} -l ${dataFile}
+          rm -Rf '${lib.head dataDirs}'/*
+          ${openldap}/bin/slapadd -F ${configDir} -n1 -l ${dataFile}
+          chown -R "${cfg.user}:${cfg.group}" ${lib.escapeShellArgs dataDirs}
         ''}
-        chown -R "${cfg.user}:${cfg.group}" "${cfg.dataDir}"
 
-        ${openldap}/bin/slaptest ${configOpts}
+        ${openldap}/bin/slaptest -u -F ${configDir}
       '';
-      serviceConfig.ExecStart =
-        "${openldap.out}/libexec/slapd -d '${cfg.logLevel}' " +
-          "-u '${cfg.user}' -g '${cfg.group}' " +
-          "-h '${concatStringsSep " " cfg.urlList}' " +
-          "${configOpts}";
-    };
-
-    users.users.openldap =
-      { name = cfg.user;
-        group = cfg.group;
-        uid = config.ids.uids.openldap;
+      serviceConfig = {
+        ExecStart = lib.concatStringsSep " " [
```

Comment:
Should be able to use `lib.escapeShellArgs` for this as well

Reply:
Done.